---
title: "Identifying Mutual Exclusivity with `TiMEx`"
author: "Simona Cristea"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Identifying Mutual Exclusivity with `TiMEx`}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Introduction
======
The TiMEx Bioconductor package is the implementation of an algoritm to identify 
groups of mutually exclusive alterations in large cancer datasets. The document
provides a detailed description of the biological datasets available with the 
TiMEx package, how to run TiMEx at full functionality and how to postprocess 
the results, as well as how to simulate data from the TiMEx model. In addition, 
an overview of the biological problem and the mathematical model of TiMEx 
are provided.

### Citation
For more details on TiMEx or for citing the TiMEx model of package, see the 
original TiMEx paper: 

*Constantinescu, S., Szczurek, E., Mohammadi, P., Rahnenfuhrer, J.,
Beerenwinkel, N.: TiMEx: A Waiting Time Model for Mutually Exclusive Cancer 
Alterations. Bioinformatics (2015)* 
[doi: 10.1093/bioinformatics/btv400](http://dx.doi.org/10.1093/bioinformatics/btv400)


The Biological Problem
-----
One of the main problems in cancer genomics is identifying the genetic 
alterations (*s.a.* point mutations or chromosomal aberrations of genes) that 
significantly contribute to tumor progression. These alterations are termed 
*drivers*, as opposed to *passengers*, alterations that have no active role
in cancer development, and simply hitchhike along with fitter clones. The most
straightforward way of separating drivers from passengers is by their frequency
alone, namely alterations which are frequent among a cohort of patients are 
identified as drivers, and all the other alterations are passengers. However, 
there are cancers for which no alteration is frequent enough so as to 
characterize cancer in a large-enough fraction of patients. Moreover, in practice, 
there are always a fraction of tumors in which tumorigenesis cannot be explained
by only the most frequent clones.

A more robust method for identifying the important genes in carcinogenesis is by
assessing their functional role in the context of groups of genes, all 
possessing the same important function, commonly known as *pathways*. Once one 
of the group members is altered, the tumor gains a significant selective 
advantage. The alteration of additional group members does not further increase 
the selective advantage of the tumor, rendering clones with a single alteration 
likely the most frequent. In this case, the group of genes displays a mutually 
exclusive alteration pattern.


The TiMEx Waiting Time Model
-----
TiMEx is a model to identify patterns of mutual exclusivity, hence leading to
identifying groups of driver genes in cancer.



## The null model

## The mutual exclusivity model

## The TiMEx procedure
TiMEx is a procedure implementing three steps: first, all pairs in the input 
dataset are tested for mutual exclusivity. Second, maximal cliques are 
identified on the basis of a selected number of pairs. Third, the resulting 
cliques are tested for mutual exclusivity.

### Step 1
In the first step, the TiMEx procedure for identifying mutually exclusive groups of alterations tests all pairs for mutual exclusivity. The data corresponding to each pair of genes is therefore fitted to both the Null (Conditional Independence) and the Mutual Exclusivity models. Parameters under the two models are estimated, and, since they are nested, a likelihood ratio test is performed between the corresponding log likelihoods, in order to test whether mu (the intensity of mutual exclusivity) is different from 0.

### Step 2

### Step 3

Running Examples
======

## Cancer datasets
The TiMEx algorithm accepts as input binary matrices, where the rows represent
patients, and the columns represent alterations. Entry `[i,j]` in the matrix is 
`1` if gene `j` was altered in patient `i`, and `0` otherwise.

The TiMEx package contains four cancer datasets: one breast cancer, one ovarian 
cancer, and two glioblastoma datasets.


Multiple datasets are available within this package. breast and ovarian are datasets downloaded from cBioPortal (TCGA) in July 2014, and preprocessed as described in Constantinescu et. al: TiMEx: A Waiting Time Model for Mutually Exclusive Cancer Alterations. Bioinformatics (2015). gbmDendrix is a glioblastoma dataset used in Leiserson et. al: Simultaneous identification of multiple driver pathways in cancer. Plos Computational Biology (2013). Additionally, this package also includes the dataset gbmMuex, used and preprocessed as described in Szczurek et. al: Modeling mutual exclusivity of cancer mutations. Research in Computational Molecular Biology (2014).


# Working examples

## Preprocessing of data

## Running TiMEx

## Running separately the three steps of the TiMEx procedure

### Step 1
The function analyzePairs performs step 1 of the TiMEx procedure: tests all gene pairs for mutual exclusivity. It returns a complex list, including parameter estimates, pairwise p-values and intensities of mutual exclusivity, log likelihoods, and others.

The output list contains exhaustive information on the pairwise testing of genes, including parameter estimates, pairwise p-values and intensities of mutual exclusivity, log likelihoods, and others.
Each element of this list is a matrix of dimension n x n (n being the number of genes), where
Element (i,j) of each matrix corresponds to the pairwise test between genes i and j. The function contains
information on the 


## Posprocessing of results

## Simulated data
Datasets can be generated from the TiMEx model using the function simulateGenes.

# Session Info 

## Styles

The `html_vignette` template includes a basic CSS theme. To override this theme you can specify your own CSS in the document metadata as follows:

    output: 
      rmarkdown::html_vignette:
        css: mystyles.css

## Figures

The figure sizes have been customised so that you can easily put two images side-by-side. 

```{r, fig.show='hold'}
plot(1:10)
plot(10:1)
```

```{r chunkName, eval=FALSE}
        myFun(3)
    ```

You can enable figure captions by `fig_caption: yes` in YAML:

    output:
      rmarkdown::html_vignette:
        fig_caption: yes

Then you can use the chunk option `fig.cap = "Your figure caption."` in **knitr**.

## More Examples

You can write math expressions, e.g. $Y = X\beta + \epsilon$, footnotes^[A footnote here.], and tables, e.g. using `knitr::kable()`.

```{r, echo=FALSE, results='asis'}
knitr::kable(head(mtcars, 10))
```

Also a quote using `>`:

> "He who gives up [code] safety for [code] speed deserves neither."
([via](https://twitter.com/hadleywickham/status/504368538874703872))
